mjpc/planners/cem/planner.cc

```diff
@@ -0,0 +1,527 @@
+                  policy.times[0],
+                  policy.times[num_spline_points - 1],
+                  timestep_power, num_spline_points);
+  }
+}
+
+// add random noise to nominal policy
+void CEMPlanner::AddNoiseToPolicy(int i) {
+  // start timer
+  auto noise_start = std::chrono::steady_clock::now();
+
+  // dimensions
+  int num_spline_points = candidate_policy[i].num_spline_points;
+  int num_parameters = candidate_policy[i].num_parameters;
+
+  // sampling token
+  absl::BitGen gen_;
+
+  // shift index
+  int shift = i * (model->nu * kMaxTrajectoryHorizon);
+
+  // sample noise
+  for (int k = 0; k < num_parameters; k++) {
+    noise[k + shift] = absl::Gaussian<double>(gen_, 0.0, stdev[k]);
+  }
+
+  // add noise
+  mju_addTo(candidate_policy[i].parameters.data(), DataAt(noise, shift),
+            num_parameters);
+
+  // clamp parameters
+  for (int t = 0; t < num_spline_points; t++) {
+    Clamp(DataAt(candidate_policy[i].parameters, t * model->nu),
+          model->actuator_ctrlrange, model->nu);
+  }
+
+  // end timer
+  IncrementAtomic(noise_compute_time, GetDuration(noise_start));
+}
+
+// compute candidate trajectories
+void CEMPlanner::Rollouts(int num_trajectory, int horizon,
+                               ThreadPool& pool) {
+  // reset noise compute time
+  noise_compute_time = 0.0;
+
+  policy.num_parameters = model->nu * policy.num_spline_points;
+
+  // random search
+  int count_before = pool.GetCount();
+  for (int i = 0; i < num_trajectory; i++) {
+    pool.Schedule([&s = *this, &model = this->model, &task = this->task,
+                   &state = this->state, &time = this->time,
+                   &mocap = this->mocap, &userdata = this->userdata, horizon,
+                   i]() {
+      // copy nominal policy
+      {
+        const std::shared_lock<std::shared_mutex> lock(s.mtx_);
+        s.candidate_policy[i].CopyFrom(s.policy, s.policy.num_spline_points);
+        s.candidate_policy[i].representation = s.policy.representation;
+      }
+
+      // sample noise policy
+      if (i != 0) s.AddNoiseToPolicy(i);
+
+      // ----- rollout sample policy ----- //
+
+      // policy
+      auto sample_policy_i = [&candidate_policy = s.candidate_policy, &i](
+                                 double* action, const double* state,
+                                 double time) {
+        candidate_policy[i].Action(action, state, time);
+      };
+
+      // policy rollout
+      s.trajectory[i].Rollout(
+          sample_policy_i, task, model, s.data_[ThreadPool::WorkerId()].get(),
+          state.data(), time, mocap.data(), userdata.data(), horizon);
+    });
+  }
+  pool.WaitCount(count_before + num_trajectory);
+  pool.ResetCount();
+}
+
+// return trajectory with best total return
+const Trajectory* CEMPlanner::BestTrajectory() {
+  return winner >= 0 ? &trajectory[winner] : nullptr;
+}
+
+// visualize planner-specific traces
+void CEMPlanner::Traces(mjvScene* scn) {
```

Comment:
need to implement some visualizations specific to CEM

Reply:
I feel like the main thing is having the purple "best" trajectory be the mean of the elite samples rather than the best sample

Reply:
This may already happen, since we set the variable `policy`'s actions as being the mean actions over the top `n_elite` samples, and apparently the purple trace is the nominal trace. See [this link](https://github.com/google-deepmind/mujoco_mpc/blob/6a56e33e37d66383a231cd5b48b26362628c3fb2/docs/OVERVIEW.md?plain=1#L190)